src/drivers/intel/cavs/dmic.c

```diff
@@ -1461,7 +1461,22 @@ static int dmic_trigger(struct dai *dai, int cmd, int direction)
 		break;
 	case COMP_TRIGGER_STOP:
 	case COMP_TRIGGER_PAUSE:
-		dmic_stop(dai);
+		if (dmic->state == COMP_STATE_ACTIVE) {
```

Comment:
I would move this to `dmic_stop` function and just exit if `dmic->state != COMP_STATE_ACTIVE`.

Reply:
I was worried about calling it in other invalid states, but it looks like dmic has just prepare/active states, so yea your solution is easier